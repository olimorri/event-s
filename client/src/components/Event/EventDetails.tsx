import React from 'react';
import { useParams } from 'react-router-dom';
import Spinner from '../Handling/Spinner';
import { Flex, Stack, Text, Image, Box, Wrap } from '@chakra-ui/react';
import Attend from './Buttons/Attend';
import Unattend from './Buttons/Unattend';
import moment from 'moment';
import { motion } from 'framer-motion';
import {Event} from '../../interfaces/Event'

type EventDetailProps = {
  events: Event[],
  signUpDown: Function,
  user: any
}

export default function EventDetails({ events, signUpDown, user }: EventDetailProps) {
  let { id }: {id: string} = useParams();
  const event: Event | undefined = events.find((el: Event) => el._id === id); //Event: {}
  const handleSubmit = (e: MouseEvent) => {
    e.preventDefault();
    signUpDown('up', id);
  };
  const unattend = (e: MouseEvent) => {
    e.preventDefault();
    signUpDown('down', id);
  };
  console.log(user.eventList);
  return (
    <Flex
      minH={'100vh'}
      justify={'center'}
      bg={'custom.100'}
      bgImage="url('https://res.cloudinary.com/dujun1hoe/image/upload/v1615228154/event-s/gradient-background-26046-26731-hd-wallpapers.jpg_cenrqe.png')"
      bgSize="100% 50%"
      backgroundRepeat="no-repeat"
    >
      {!event ? (
        <Spinner />
      ) : (
        <Box
          w={'60%'}
          minW={'60%'}
          maxW={'60%'}
          bg={'gray.50'}
          marginTop={50}
          borderRadius="md"
          boxShadow={'md'}
          m={20}
          h={'60vw'}
          maxH={'80vw'}
        >
          <motion.div
            initial={{ opacity: 0 }}
            animate={{
              opacity: 1,
              transition: { delay: 0.4 },
            }}
          >
            <Wrap w={'100%'} align={'center'} p={1}>
              <Box w={'50%'} minW={'300px'}>
                <Image
                  borderRadius="md"
                  boxShadow="md"
                  src={event.photo}
                  alt=""
                />
              </Box>
              <Stack w={'50%'} align={'center'}>
                <Flex flexDirection={'column'} justify={'center'}>
                  <Text fontSize="3xl" m={3}>
                    {event.name}
                  </Text>
                  <Text align={'center'} fontSize="xl">
                    {event.location}
                  </Text>
                </Flex>
                {!Object.keys(user).length ||
                user.host ? null : !user.eventList.some(
                    (el: Event) => el._id === id,
                  ) ? (
                  <Attend handleSubmit={handleSubmit} />
                ) : (
                  <Unattend unattend={unattend} />
                )}
              </Stack>
            </Wrap>
            <Box p={4}>
              <Stack
                w={'100%'}
                h={'60vh'}
                direction={{ base: 'column', md: 'row' }}
              >
                <Flex
                  h={'60%'}
                  flex={1}
                  pl={4}
                  align={'end'}
                  justifyContent={'space-evenly'}
                  flexDirection={'column'}
                >
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.5,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/calendar-2_hmjf1m.png'
                        }
                        alt=""
                      />
                      <Text justify={'center'}>
                        {moment(event.date).format('MMM Do YY')}
                      </Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.6,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296946/event-s/icons/alarm-2_csdsdf.png'
                        }
                        alt=""
                      />
                      <Text>{event.duration}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 0.7,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615297028/event-s/icons/announcement_pr7pea.png'
                        }
                        alt=""
                      />
                      <Text>{event.limit}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.8,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/event_h6dtmv.png'
                        }
                        alt=""
                      />
                      <Text>{event.type}</Text>
                    </Wrap>
                  </motion.div>
                  <motion.div
                    initial={{
                      opacity: 0,
                      x: '-100vw',
                    }}
                    animate={{
                      opacity: 1,
                      x: 0,
                    }}
                    transition={{
                      // type: 'spring',
                      // @ts-ignore
                      ease: 'easeIn',
                      delay: 0.9,
                      duration: 0.5,
                    }}
                  >
                    <Wrap align={'center'}>
                      <Image
                        borderRadius="md"
                        src={
                          'https://res.cloudinary.com/dujun1hoe/image/upload/v1615296945/event-s/icons/done_r4quc3.png'
                        }
                        alt=""
                      />
                      <Text>{event.attendees}</Text>
                    </Wrap>
                  </motion.div>
                </Flex>
                <Flex
                  h={'100%'}
                  flex={2}
                  p={4}
                  alignContent={'center'}
                  flexDirection={'column'}
                >
                  <Text fontWeight="bold" align={'center'}>
                    About this event
                  </Text>
                  <motion.div
                    initial={{
                      opacity: 0,
                      scale: 0,
                    }}
                    animate={{
                      opacity: 1,
                      scale: 1,
                    }}
                    transition={{
                      // type: 'spring',
                      ease: 'easeIn',
                      delay: 1,
                      duration: 1,
                    }}
                  >
                    <Text mt={10} align={'center'}>
                      {event.description.slice(0, 1200)}...
                    </Text>
                  </motion.div>
                </Flex>

                <Stack spacing={4}></Stack>
              </Stack>
            </Box>
          </motion.div>
        </Box>
      )}
    </Flex>
  );
}
