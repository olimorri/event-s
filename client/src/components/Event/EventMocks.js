const events = [
  {
    limit: 1,
    duration: 1,
    attendees: -2,
    list: ['604a012800da7a8522af8b20'],
    _id: '604a268300da7a8522af8b23',
    name: 'Test Event',
    location: 'SE14 5NP',
    type: 'Food',
    description: 'sdfsdfew',
    date: '2021-03-20T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    geolocation: '-0.042843,51.470509',
    owner: '604a0b6300da7a8522af8b21',
    __v: 0,
  },
  {
    limit: 10000000,
    duration: 1,
    attendees: 0,
    list: [],
    _id: '604ce151b61ff67613b5b6d9',
    geolocation: '-0.155083,50.840887',
    owner: '604a0b6300da7a8522af8b21',
    name: 'Another Test',
    location: 'BN1 5AN',
    type: 'Food',
    description: 'Test',
    date: '2021-03-20T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 100,
    duration: 1,
    attendees: 0,
    list: [],
    _id: '604ce171b61ff67613b5b6da',
    geolocation: '-0.155083,50.840887',
    owner: '604a0b6300da7a8522af8b21',
    name: 'Another Test',
    location: 'BN1 5AN',
    type: 'Food',
    description: 'dgfgdfgdfgf',
    date: '2021-03-20T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 12,
    duration: 1,
    attendees: 0,
    list: [],
    _id: '604ce3d6b61ff67613b5b6db',
    geolocation: '0.110623,50.767652',
    owner: '604a0b6300da7a8522af8b21',
    name: 'eatrhdfghgfhfg',
    location: 'bn251az',
    type: 'Food',
    description: 'dfghdgfhgfdh',
    date: '2021-04-03T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 12,
    duration: 12,
    attendees: -15,
    list: ['604a012800da7a8522af8b20'],
    _id: '604f8bd4cdb45af7cd1ca265',
    geolocation: '-0.042843,51.470509',
    owner: '604a0b6300da7a8522af8b21',
    name: '12312323213123',
    location: 'se14 5np',
    type: 'Leisure',
    description: 'dfgdfsgdfg',
    date: '2021-03-31T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 221,
    duration: 22,
    attendees: -1,
    list: [],
    _id: '604f8c11cdb45af7cd1ca266',
    geolocation: '0.110623,50.767652',
    owner: '604a0b6300da7a8522af8b21',
    name: '99999999',
    location: 'bn251az',
    type: 'Sport',
    description: 'dfgdfggdf',
    date: '2021-03-28T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 2,
    duration: 1,
    attendees: 0,
    list: [],
    _id: '604f8c49cdb45af7cd1ca267',
    geolocation: '-3.597458,50.562947',
    owner: '604a0b6300da7a8522af8b21',
    name: 'Motherfucking Test',
    location: 'tq123pf',
    type: 'Music',
    description: 'sdfsdfdsf',
    date: '2021-03-26T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
  {
    limit: 3,
    duration: 1,
    attendees: 0,
    list: [],
    _id: '604fa49085cdd6032dfdf5e0',
    geolocation: '-0.049937,51.470627',
    owner: '604a0b6300da7a8522af8b21',
    name: 'The Final Test',
    location: 'SE14 5ND',
    type: 'Sport',
    description: 'sdfrsdfddf',
    date: '2021-03-31T00:00:00.000Z',
    photo: 'https://source.unsplash.com/random/300x300',
    __v: 0,
  },
];

const event = {
  attendees: 1,
  date: '2021-03-23T00:00:00.000Z',
  description: 'This is a test event',
  duration: 2,
  geolocation: '0.261773,51.116312',
  limit: 2,
  list: ['604a36891b95d669d99ebe2e'],
  location: 'tn25hs',
  name: 'test event',
  owner: '604b5c00ca29f24f751f492c',
  photo: 'https://source.unsplash.com/random/300x300',
  type: 'Food',
  _id: '604b5c00ca29f24f751f492c',
};

module.exports = { events, event };
