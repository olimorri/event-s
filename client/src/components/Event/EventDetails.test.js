import {
  render,
  cleanup,
  screen,
  waitForElementToBeRemoved,
  queryByText,
  findByText,
  getByText,
  wait,
  waitFor,
  configure,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import EventDetails from './EventDetails';
import { events, event } from './EventMocks';

afterEach(cleanup);

configure({
  asyncUtilTimeout: 5000,
});

const EventDetailsTest = () => (
  <BrowserRouter>
    <EventDetails events={events} event={event} />
  </BrowserRouter>
);

afterEach(cleanup);

// let parent;

// beforeAll(() => (parent = render(<EventDetailsTest />)));

describe('Snapshot', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<EventDetailsTest />);
    expect(asFragment(<EventDetailsTest />)).toMatchSnapshot();
  });
});

describe('Unit Testing', () => {
  // test('renders img', () => {
  //   render(<EventDetailsTest />);
  //   const displayedImg = document.querySelector('img');
  //   expect(displayedImg).toBeValid();
  // });

  test('rendered event name matches props event.name', async () => {
    render(<EventDetailsTest />);

    const loading = screen.getByText('Loading...');
    expect(loading).toHaveTextContent(/loading/i);

    await waitFor(() => {
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    });
    // const eventName = screen.queryByText(event.name);
    // console.log(eventName);
    // expect(eventName).toHaveTextContent('test event');
  });

  // test('rendered event postcode matches props event.location', () => {
  //   render(<EventDetailsTest />);
  //   const eventPostcode = screen.getByText(event.location);
  //   expect(eventPostcode).toBeInTheDocument();
  // });

  // test('rendered event date matches props event.date', () => {
  //   render(<EventDetailsTest />);
  //   const eventDate = screen.getByText(/Mar 23rd 21/i);
  //   expect(eventDate).toBeInTheDocument();
  // });
});
