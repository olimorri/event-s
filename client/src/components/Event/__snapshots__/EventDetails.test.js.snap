// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1pv4q05"
  >
    <div
      class="css-2j8non"
    >
      <div
        class="loader"
      >
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
